### PR TITLE
[Relay] Support i16, f16 scalars in Relay text

### PR DESCRIPTION
**Pull request by mbs-octoml**
While testing fp16 models for Collage discovered the Relay text
format did not support f16. While adding that cleaned up scalar handling
in general. However I left two inlined tests for 'is simple const'
in place (fuse_ops.cc and memory_alloc.cc) since it's not clear whether
they should remain specific to just {i,f}{32,64} or whether they can
be replaced with the support::IsSimpleScalar central predicate.

Also include spans in pretty printed form since I've started populating
them when debugging Collage.

### CODE REVIEW

**Review by Mousius — CHANGES_REQUESTED** [review]
I think this is generally an improvement, but I'm curious whether we could cover more of the error scenarios in test cases? There's three different diagnostic errors for both int and float plus the clipping errors - though I'm unsure how easy it is to trigger them with current testing infrastructure.

File: tests/python/relay/test_ir_text_printer.py [hunk 1]
```diff
@@ -47,16 +47,28 @@ def show(text):
...
+def test_scalars():
+    assert_prints_as(relay.const(42, "int16"), "42i16")
+    assert_prints_as(relay.const(42, "int32"), "42")
+    assert_prints_as(relay.const(42, "int64"), "42i64")
+    assert_prints_as(relay.const(3.0, "float16"), "3f16")
+    assert_prints_as(relay.const(3.0, "float32"), "3f")
+    assert_prints_as(relay.const(3.0, "float64"), "3f64")
+
+
+def test_large_graph():
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply tests/python/relay/test_ir_text_printer.py:63]: Yeah, turned it on expecting CI to fail but all was well. Job well done indeed.

**Comment by mbs-octoml** [comment]
Thanks @Mousius (and sorry for the delay, been off an a separate adventure). I think the range checking testing is ok, and I added range checks for fp16 too. For the LOG(FATALS) I'll add cc tests since it's tricky to tickle the failures via python. stand by.

**Review by Mousius — APPROVED** [review]

**Comment by Mousius** [comment]
Thanks @mbs-octoml 😸  nice clean up!

**Pull request merged**